src/shared/dissect-image.c

```diff
@@ -2241,28 +2245,38 @@ int dissected_image_mount(
                 const char *esp_path = NULL;
 
                 if (where) {
+                        bool missing_ok = false;
+
                         /* Mount the ESP to /boot/ if it exists and is empty and we didn't already mount the
-                         * XBOOTLDR partition into it. Otherwise, use /efi instead, but only if it exists
-                         * and is empty. */
+                         * XBOOTLDR partition into it. Otherwise, try with /efi and /boot/efi (in that
+                         * order). If that fails it's likely none of the previous paths exist. Try again but
+                         * don't make the existence of the directory mandatory this time, the directory (/efi
+                         * by default) will be created when mounting ESP. If that fails again then the
+                         * directories exist but are not empty. */
+again:
```

Comment:
coding style: goto is only ok for cleanup paths. never for loops, i.e. never to jump ip again

Reply:
the goto was used to emphasize that looping again was not something we really expect (ie not normal execution) but ok noted.